eth/sync.go

```diff
@@ -185,9 +185,15 @@ func (pm *ProtocolManager) synchronise(peer *peer) {
 		// The only scenario where this can happen is if the user manually (or via a
 		// bad block) rolled back a fast sync node below the sync point. In this case
 		// however it's safe to reenable fast sync.
+
+		// Make sure the peer's total difficulty we are synchronizing is higher.
+		if pm.blockchain.GetTdByHash(pm.blockchain.CurrentFastBlock().Hash()).Cmp(pTd) >= 0 {
+			return
+		}
```

Comment:
This is only one path where fast sync is enabled, the above if branch also enables fast sync.

Please move these two lines below outside of both branches, and do something like

```go
if mode == downloader.FastSync {
	if pm.blockchain.GetTdByHash(pm.blockchain.CurrentFastBlock().Hash()).Cmp(pTd) >= 0 {
		return
	}
}
```